File: 72414_2.swift
import Foundation

func toSeconds(_ time: String) -> Int {
    let splited = time.components(separatedBy: ":")
    guard let hours = Int(splited[0]), let minutes = Int(splited[1]), let seconds = Int(splited[2]) else { return 0 }
    return hours * 3600 + minutes * 60 + seconds
}

func toString(_ time: Int) -> String {
    let hours = time / 3600
    let minutes = (time - hours * 3600) / 60
    let seconds = time - hours * 3600 - minutes * 60
    return String(format: "%02d:%02d:%02d", hours < 0 ? 0 : hours, minutes < 0 ? 0 : minutes, seconds < 0 ? 0 : seconds)
}

func solution(_ play_time:String, _ adv_time:String, _ logs:[String]) -> String {
    let playTime = toSeconds(play_time)
    var adTime = toSeconds(adv_time)
    var accumulatedViews = Array(repeating: 0, count: playTime + 1)
    var candidatePoints: [Int] = []
    
    for log in logs {
        let splited = log.components(separatedBy: "-")
        let from = toSeconds(splited[0])
        let to = toSeconds(splited[1])
        
        accumulatedViews[from] += 1
        accumulatedViews[to] -= 1
        
        candidatePoints.append(from)
        candidatePoints.append(to - adTime)
    }

    for i in 1..<playTime { accumulatedViews[i] += accumulatedViews[i - 1] }
    for i in 1..<playTime { accumulatedViews[i] += accumulatedViews[i - 1] }
    
    var bestTime = 0 
    var bestViews = accumulatedViews[adTime - 1]
    
    for candidate in candidatePoints {
        let currentViews = accumulatedViews[min(candidate + adTime - 1, accumulatedViews.count - 1)] - (candidate - 1 > 0 ? accumulatedViews[candidate - 1] : 0)
        if currentViews > bestViews || (currentViews == bestViews && bestTime > candidate) {
            bestViews = currentViews
            bestTime = candidate
        }
    }
    
    return toString(bestTime)
}

File: 72414_2.cpp
#include <string>
#include <vector>
#include <iostream>
#include <queue>

using namespace std;

int stringTimeToInt(string time) {
    return stoi(string(time.begin(), time.begin() + 2)) * 3600 + stoi(string(time.begin() + 3, time.begin() + 5)) * 60 + stoi(string(time.begin() + 6, time.begin() + 8));
}

string intTimeToString(int time) {
    char temp[9];
    int hours = time / 3600;
    int minutes = (time - hours * 3600) / 60;
    int seconds = (time - hours * 3600) - minutes * 60;
    sprintf(temp, "%02d:%02d:%02d", hours, minutes, seconds);
    return temp;
}

bool compareCandidates(int left, int right) {
    return left < right;
}

string solution(string play_time, string adv_time, vector<string> logs) {
    int endTime = stringTimeToInt(play_time);
    int adDuration = stringTimeToInt(adv_time);
    vector<int> viewers = vector<int>(endTime + 1, 0);
    
    auto compareCandidates = [](int left, int right) { return left > right; };
    priority_queue<int, vector<int>, decltype(compareCandidates)> candidates(compareCandidates);
    candidates.push(0);
    
    for(auto log: logs) {
        int starting = stringTimeToInt(string(log.begin(), log.begin() + 8));
        int ending = stringTimeToInt(string(log.begin() + 9, log.begin() + 17));
        for(int playTime = starting; playTime < ending; playTime++) { viewers[playTime] += 1; }
    }
    
    queue<int> views;
    long long bestViewers = 0;
    long long currentViewers = 0;
    int from = 0;
    int to = adDuration;
    int bestTime = 0;
    
    for(int i = from; i < to; i++) {
        currentViewers += viewers[i];
        views.push(viewers[i]);
    }
    bestViewers = currentViewers;
    
    while(to <= endTime) {
        currentViewers += viewers[to];
        currentViewers -= viewers[from];
        views.push(viewers[to]);
        views.pop();
        
        from++;
        to++;
        
        if(currentViewers > bestViewers) {
            bestViewers = currentViewers;
            bestTime = from;
        }
    }
    
    return intTimeToString(bestTime);
}
